Add styles

octopusinvitro opened the pull request:
Visit the site here:

https://donation-system-staging.herokuapp.com/

In the designs, there is a background image that has to cover
the main top area and whose bottom has to be aligned with the
top of  the gift aid area. Also, the main quote had to be
contained in  the top area, in the limits of the background
image.

The layout that worked is one in which the form is the main
container, with a page-top and a page-bottom areas.

* The top area has the form elements that go before the gift
aid area, and the quote.

* The bottom area has the gift aid area followed by the rest of
the form elements, and the information links

I also kept in mind that the final stylesheet size should not be over
around 20k, since there are also custom fonts, images and the
payment-related javascript libraries, that need to be downloaded.

I didn't have designs for mobile, so I made them up.

The fonts and images are served from Amazon S3.

Another thing that has to be done at some point is to add Autoprefixer.
I added an issue for that #32 

The layout is as follows:
```

--------------------------------------------
|  form as a container                     |
|  --------------------------------------  |
|  |  page-top (with bg image)          |  |
|  |   ______________   _____________   |  |
|  |  |              |  |            |  |  |
|  |  |  top inputs  |  |   quote    |  |  |
|  |  |______________|  |____________|  |  |
|  |                                    |  |
|  |------------------------------------|  |
|  |  page-bottom                       |  |
|  |   ______________   _____________   |  |
|  |  |              |  |            |  |  |
|  |  |  gift area   |  |   info     |  |  |
|  |  |______________|  |____________|  |  |
|  |  |              |                  |  |
|  |  | other inputs |                  |  |
|  |  |______________|                  |  |
|  |                                    |  |
|  --------------------------------------  |
--------------------------------------------
```

## Screenshots

Home page, mobile

![mobile](https://user-images.githubusercontent.com/2157089/35475034-77b37420-038f-11e8-97af-4e6183955a5e.png)

Home page, desktop

![desktop](https://user-images.githubusercontent.com/2157089/35475035-7e9cab94-038f-11e8-9630-7fa560c9de86.jpg)

Success page, mobile

![mobile-success](https://user-images.githubusercontent.com/2157089/35475043-899dd838-038f-11e8-88ab-d9e1bf4e99f3.png)

Success page, desktop

![desktop-success](https://user-images.githubusercontent.com/2157089/35475051-91390748-038f-11e8-8ab7-6254ce783b2c.jpg)

## Performance, accessibility, etc.

I passed the resulting styled page through some audit tools, although these are not exhaustive. I'll look into how to fix some things like gzipping text files in Heroku, etc.

### GTMetrix: 

Almost 3s and 730KB. It's not ideal, but the custom fonts are heavy and there are a lot of javascript libraries related with the payment that account for most of the 29 requests listed

![gtmetrix-audit](https://user-images.githubusercontent.com/2157089/34005519-02faff5e-e0f3-11e7-815d-42397620168f.png)

### Lighthouse:

Complains about not serving stuff through HTTP2, which is something I'll have to look into. I'll fix the progressive web app complains in another PR

![lighthouse-audit](https://user-images.githubusercontent.com/2157089/34005622-6a74f1d0-e0f3-11e7-9746-1bd92b2870aa.png)

### Wave (accessibility)

No errors and the two alerts that lighthouse mentioned, related the lack of contrast of some elements. I can't really fix this as that is the design I was given

![wave-audit](https://user-images.githubusercontent.com/2157089/34005697-aed9078a-e0f3-11e7-92a8-efaecfe4184c.png)

Waterfall:

![damn-fonts](https://user-images.githubusercontent.com/2157089/34006228-802a179c-e0f5-11e7-89c9-05b6365a7263.png)

octopusinvitro committed:
Add main variables, fonts and colors to main

octopusinvitro committed:
Style pages according to contractor designs

In the designs, there is a background image that has to cover
the main top area and whose bottom has to be aligned with the
top of  the gift aid area. Also, the main quote had to be
contained in  the top area, in the limits of the background
image.

I'm not a designer, so after several unsuccessful approaches,
the layout that worked is one in which the form is the main
container, with a page-top and a page-bottom areas.

* The top area has the form elements that go before the gift
aid area, and the quote.

* Tho bottom area has the gift aid area followed by the rest of
the form elements, and the information links

I also kept in mind that the final file size should not be over
around 20k, since there are also custom fonts, images and the
payment-related javascript libraries, that need to be downloaded.

The layout is as follows:

--------------------------------------------
|  form as a container                     |
|  --------------------------------------  |
|  |  page-top (with bg image)          |  |
|  |   ______________   _____________   |  |
|  |  |              |  |            |  |  |
|  |  |  top inputs  |  |   quote    |  |  |
|  |  |______________|  |____________|  |  |
|  |                                    |  |
|  |------------------------------------|  |
|  |  page-bottom                       |  |
|  |   ______________   _____________   |  |
|  |  |              |  |            |  |  |
|  |  |  gift area   |  |   info     |  |  |
|  |  |______________|  |____________|  |  |
|  |  |              |                  |  |
|  |  | other inputs |                  |  |
|  |  |______________|                  |  |
|  |                                    |  |
|  --------------------------------------  |
--------------------------------------------

octopusinvitro committed:
Update README

ketan-survival commented:
👍  for benchmarks

ketan-survival commented on lib/donation_system_webapp.rb:
```diff
@@ -19,12 +19,12 @@ class DonationSystemWebapp < Sinatra::Base
   post '/donations' do
     errors = Donations.donate(params)
-    pass if errors.empty?
+    redirect '/success' if errors.empty?
```
That reads a lot better to me.

ketan-survival reviewed (approved):
LGTM

octopusinvitro committed:
Add the Amazon gem

octopusinvitro committed:
Add a class to upload timestamped files to Amazon S3

This uses the AWS SDK gem to upload CSS and JS to the staging
and production buckets.

octopusinvitro committed:
Disable RSpec warnings

The warnings were disabled in the spec helper because the aws-sdk
gem is throwing loads of warnings from its different subgems.

octopusinvitro committed:
Install the Heroku API gem

This gem will allow to set configuration variables from the app.

octopusinvitro committed:
Add a Heroku client

The Heroku client uploads the new timestamped filenames of the
CSS and JS assets in an upstream-only environment variable.

This allows to synchronize the last filename uploaded to AWS S3
with the path shown in the rendered HTML at Heroku servers.

The client also pushes the local app to Heroku using git push.
There is a way to do this using the API and the gem, but it is
too complicated and with git you can do in in a on-line command.

octopusinvitro committed:
Add a very basic error printer for the CLI

When the app is run in the terminal, the errors accumulated
during execution will be printed to the screen thanks to this
class.

octopusinvitro committed:
Add a deployer class to control all the classes

octopusinvitro committed:
Add a class to deal with appending a timestamp

The reason to append a timestamp to styles and JavaScripts is to
force browsers to recache these files when they have changed in
staging or in production.

This class also generates both the local and upstream paths to
the datestamped files, so that they can be used in other deploy
files.

octopusinvitro committed:
Add a Rake task to deploy to staging or production

Since now we have a rake tast to upload the static assets to S3
and deploy, there is no need to keep the main.css and bundle.js
in the repo, hence they were added to the .gitignore

octopusinvitro committed:
Add a class to generate the right paths for assets

This class generates the paths that go in the views, to load
the main.css file or the bundler.js

octopusinvitro committed:
Use the asset path generator in the views

octopusinvitro committed:
Update README

octopusinvitro committed:
Add further explanations for beginners, designers, etc.

These things may be obvious to us, but newcomers, beginners or
designers wanting to work on the frontend of this app may not
have it that easy working with a Ruby app.

octopusinvitro committed:
Add a section in the README explaining how to update things
